src/NServiceBus.RavenDB/SessionManagement/OpenRavenSessionByCustomDelegate.cs

```diff
@@ -6,20 +6,22 @@
 
     class OpenRavenSessionByCustomDelegate : IOpenTenantAwareRavenSessions
     {
-        public OpenRavenSessionByCustomDelegate(Func<IDictionary<string, string>, IAsyncDocumentSession> getAsyncSession)
+        public OpenRavenSessionByCustomDelegate(Func<IDictionary<string, string>, IAsyncDocumentSession> getAsyncSession, SagaPersistenceConfiguration sagaPersistenceConfiguration)
         {
             getAsyncSessionUsingHeaders = getAsyncSession;
+            this.sagaPersistenceConfiguration = sagaPersistenceConfiguration;
         }
 
         public IAsyncDocumentSession OpenSession(IDictionary<string, string> messageHeaders)
         {
             var session = getAsyncSessionUsingHeaders(messageHeaders);
 
-            //session.Advanced.UseOptimisticConcurrency = true;
+            session.Advanced.UseOptimisticConcurrency = !sagaPersistenceConfiguration.EnablePessimisticLocking;
```

Comment:
Shouldn't this always be `true`? The pessimistic locking applies only when all parties 'play along'. Also, if a customer would be using custom RavenDB IO from a handler this handler would need to share the same session that might still require OCC. I would leave this to true as a fail-safe for ourselves and also for customers.

Maybe this can be controlled per document in a session? In <https://ravendb.net/docs/article-page/4.2/csharp/client-api/session/configuration/how-to-enable-optimistic-concurrency> I read that is can only be disabled for existing documents and enabled for new documents but not disabled for existing.

Reply:
You’re right, fixed. I was confused previously by failing tests.

We cannot really optimize on a document basis as it requires a change vector to be set for every saved document. It might also have implications on how users store their documents.